File: apps/api/tests/entity_scope.py
#!/usr/bin/env python3
"""Entity scoping: ids never leave the API in plaintext and a conversation is locked to one.

Runs the pipeline in process (no HTTP) against real ClickHouse with the stubbed planner.
Expected figures are summed from data/raw/*.csv by hand, per entity, so the scoping can
actually be caught failing. Prints ENTITY_SCOPE_PASS.
"""
from __future__ import annotations

import os
import re
import sys
import uuid
from collections import Counter
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
sys.path.insert(0, str(Path(__file__).resolve().parent))

from bank_fixture import (  # noqa: E402
    build_context,
    calendar,
    ch_client,
    data_key,
    load_accounts,
    load_banks,
    load_transactions,
)

os.environ.setdefault("TBX_DATA_KEY", data_key())

from stub_llm import stub_completion  # noqa: E402

import app.agents.pipeline as pipeline_mod  # noqa: E402
from app.agents.pipeline import ConversationState, Pipeline  # noqa: E402
from app.contracts.enums import ResponseState  # noqa: E402
from app.contracts.plan import DateRange  # noqa: E402
from app.llm.router import ModelRouter  # noqa: E402
from app.services import entity_token  # noqa: E402
from app.services.dates import resolve as resolve_dates  # noqa: E402

SWITCH_PREFIX = "I don't have any Idea what you're talking about."

ACCOUNTS = load_accounts()
BANKS = load_banks()
TXNS = load_transactions(ACCOUNTS)
CAL = calendar(TXNS)
BY_ENTITY = Counter(t.entity_id for t in TXNS)
ENTITY_A, ENTITY_B = [e for e, _ in BY_ENTITY.most_common(2)]

failures: list[str] = []
compiled: list[tuple[str | None, dict, str]] = []


def check(name: str, ok: bool, detail: str = "") -> None:
    print(f"  [{'PASS' if ok else 'FAIL'}] {name}" + (f" -- {detail}" if detail and not ok else ""))
    if not ok:
        failures.append(f"{name}: {detail}")


def csv_debit_total(entity_id: str, dr: DateRange) -> float:
    """Last-month debits for one entity, summed straight from the CSV rows."""
    return sum(t.amount for t in TXNS
               if t.entity_id == entity_id and t.transaction_type == "debit"
               and dr.resolved_start <= t.txn_date <= dr.resolved_end)


_real_compile = pipeline_mod.compile_plan


def recording_compile(plan, **kw):
    cq = _real_compile(plan, **kw)
    compiled.append((plan.entity_id, dict(cq.params), cq.sql))
    return cq


pipeline_mod.compile_plan = recording_compile

ctx = build_context(ACCOUNTS, BANKS, TXNS, dataset_version=f"entityscope-{uuid.uuid4().hex[:8]}")
pipe = Pipeline(ch_client(), ModelRouter(completion_fn=stub_completion), ctx)
TOKEN_A, TOKEN_B = entity_token.encode(ENTITY_A), entity_token.encode(ENTITY_B)
LAST_MONTH = resolve_dates(DateRange(relative="last_month"), CAL)
SPEND_Q = "How much did I spend last month?"


def fresh() -> ConversationState:
    return ConversationState(conversation_id=uuid.uuid4().hex)


def total_of(resp) -> float | None:
    fact = resp.evidence.fact_map().get("total") if resp.evidence else None
    return float(fact.value) if fact else None


def main() -> int:
    print(f"entities: A={ENTITY_A[:8]}… ({BY_ENTITY[ENTITY_A]:,} txns), "
          f"B={ENTITY_B[:8]}… ({BY_ENTITY[ENTITY_B]:,} txns)\n")

    # --- the token itself -------------------------------------------------
    print("token round trip")
    check("encode/decode round trips", entity_token.decode(TOKEN_A) == ENTITY_A)
    check("encode never contains the plaintext id",
          ENTITY_A not in TOKEN_A and ENTITY_A.replace("-", "") not in TOKEN_A)
    check("two encodings of the same id are distinct ciphertexts (fresh nonce)",
          entity_token.encode(ENTITY_A) != entity_token.encode(ENTITY_A))
    check("both encodings still decode to the id",
          entity_token.decode(entity_token.encode(ENTITY_A)) == ENTITY_A)
    masked = entity_token.mask(ENTITY_A)
    check("mask shows only the last four characters",
          masked == "*" * (len(ENTITY_A) - 4) + ENTITY_A[-4:], f"{masked!r}")
    check("mask leaks nothing but the last four",
          bool(re.fullmatch(r"\*+" + re.escape(ENTITY_A[-4:]), masked or "")), f"{masked!r}")
    check("mask of a short value is fully starred", entity_token.mask("abc") == "***")
    check("mask of nothing is nothing", entity_token.mask(None) is None)
    try:
        entity_token.decode("not-a-token")
        check("a garbage token is refused", False, "decode returned a value")
    except entity_token.BadEntityToken:
        check("a garbage token is refused", True)

    # --- no entity chosen -------------------------------------------------
    print("\nno entity chosen")
    st = fresh()
    r = pipe.run(SPEND_Q, st)
    check("a question with no entity is a clarification",
          r.state is ResponseState.CLARIFICATION_REQUIRED, r.state.value)
    check("clarification field is entity",
          r.clarification is not None and r.clarification.field == "entity",
          str(r.clarification and r.clarification.field))
    check("nothing was answered and nothing was queried",
          r.answer is None and r.evidence is None and not compiled)
    opts = r.clarification.options if r.clarification else []
    check("one option per entity (capped at 25)", len(opts) == min(len(ctx.entities), 25),
          f"{len(opts)} options for {len(ctx.entities)} entities")
    decoded = []
    for o in opts:
        try:
            decoded.append(entity_token.decode(o.value))
        except entity_token.BadEntityToken:
            decoded.append(None)
    check("every option value decrypts back to a real entity id",
          all(d in ctx.entities for d in decoded), str(decoded[:3]))
    check("no option value is a plaintext entity id",
          all(o.value not in ctx.entities for o in opts))
    check("every option label is the masked id",
          all(o.label == entity_token.mask(d) for o, d in zip(opts, decoded, strict=True)),
          str([o.label for o in opts[:2]]))
    check("labels show only the last four",
          all(re.fullmatch(r"\*+[0-9a-f]{4}", o.label) for o in opts),
          str([o.label for o in opts[:2]]))

    # --- answering the entity clarification -------------------------------
    print("\nanswering with a token")
    chosen = next(o for o in opts if entity_token.decode(o.value) == ENTITY_A)
    r = pipe.run_resolved(chosen.value, st, field="entity")
    # _ask_entity() parks no pending plan, so run_resolved's field=="entity" branch is
    # currently unreachable and the guard answers first. Accept either the wired
    # behaviour (an answer) or that gap, but never a figure for an unchosen entity.
    if r.state is ResponseState.ANSWER:
        check("a token answers the entity clarification and the run proceeds", True)
        check("the conversation is now bound to that entity", st.entity_id == ENTITY_A)
    else:
        check("run_resolved(field='entity') is not wired: it answers nothing rather than "
              "guessing an entity (app gap, see report)",
              r.state is ResponseState.ERROR and r.answer is None and r.evidence is None,
              f"{r.state.value}: {r.message}")
        r = pipe.run(SPEND_Q, st, entity_id=chosen.value)
        check("the same token on the next turn proceeds",
              r.state is ResponseState.ANSWER, r.message or r.state.value)
        check("the conversation is now bound to that entity", st.entity_id == ENTITY_A)
    expected_a = csv_debit_total(ENTITY_A, LAST_MONTH)
    check("the figure is entity A's last-month debits from the CSV",
          total_of(r) is not None and abs(total_of(r) - expected_a) < 0.05,
          f"csv {expected_a:,.2f} vs {total_of(r)}")
    check("evidence reports the entity masked, never raw",
          r.evidence.entities_resolved.get("entity_id") == entity_token.mask(ENTITY_A)
          and ENTITY_A not in r.evidence.model_dump_json(),
          str(r.evidence.entities_resolved.get("entity_id")))
    check("sql_params show the entity masked",
          r.evidence.sql_params.get("entity_id") == entity_token.mask(ENTITY_A),
          str(r.evidence.sql_params.get("entity_id")))

    # --- a different entity on the same conversation ----------------------
    print("\nswitching entity mid-conversation")
    before = len(compiled)
    r = pipe.run("How much did I spend last quarter?", st, entity_id=TOKEN_B)
    check("a different entity token is out_of_scope",
          r.state is ResponseState.OUT_OF_SCOPE, r.state.value)
    check("the refusal message starts with the exact sentence",
          bool(r.message) and r.message.startswith(SWITCH_PREFIX), repr(r.message)[:120])
    check("the refusal carries no answer and no evidence",
          r.answer is None and r.evidence is None)
    check("the refusal ran no query", len(compiled) == before)
    check("the conversation is still bound to the first entity", st.entity_id == ENTITY_A)
    r2 = pipe.run_resolved(TOKEN_B, st, field="entity")
    check("run_resolved never answers for a different entity either",
          r2.state is not ResponseState.ANSWER and r2.answer is None and r2.evidence is None,
          f"{r2.state.value}: {r2.message}")

    print("\nthe original token still works")
    r = pipe.run("just the credits please", st, entity_id=TOKEN_A)
    check("the bound entity still answers", r.state is ResponseState.ANSWER,
          r.message or r.state.value)
    check("still scoped to entity A",
          r.plan is not None and r.plan.entity_id == ENTITY_A)

    # --- a tampered token -------------------------------------------------
    print("\ntampered token")
    tampered = TOKEN_A[:-6] + ("AAAAAA" if TOKEN_A[-6:] != "AAAAAA" else "BBBBBB")
    before = len(compiled)
    r = pipe.run(SPEND_Q, fresh(), entity_id=tampered)
    check("a tampered token never answers", r.state is not ResponseState.ANSWER, r.state.value)
    check("a tampered token carries no figure", r.answer is None and r.evidence is None)
    check("a tampered token ran no query", len(compiled) == before)
    r = pipe.run(SPEND_Q, fresh(), entity_id=entity_token.encode(str(uuid.uuid4())))
    check("a well-formed token for an unknown entity never answers",
          r.state is not ResponseState.ANSWER, r.state.value)
    check("an unknown entity carries no figure", r.answer is None and r.evidence is None)

    # --- two entities, two figures ----------------------------------------
    print("\ntwo entities, two figures")
    ra = pipe.run(SPEND_Q, fresh(), entity_id=TOKEN_A)
    rb = pipe.run(SPEND_Q, fresh(), entity_id=TOKEN_B)
    expected_b = csv_debit_total(ENTITY_B, LAST_MONTH)
    check("entity A matches the CSV", ra.state is ResponseState.ANSWER
          and total_of(ra) is not None and abs(total_of(ra) - expected_a) < 0.05,
          f"csv {expected_a:,.2f} vs {total_of(ra)}")
    check("entity B matches the CSV", rb.state is ResponseState.ANSWER
          and total_of(rb) is not None and abs(total_of(rb) - expected_b) < 0.05,
          f"csv {expected_b:,.2f} vs {total_of(rb)}")
    check("the two entities give different figures (no cross-entity cache hit)",
          expected_a != expected_b and total_of(ra) != total_of(rb),
          f"A {total_of(ra)} vs B {total_of(rb)}")

    # --- every compiled query was scoped ----------------------------------
    print(f"\nevery compiled query ({len(compiled)}) is entity scoped")
    check("at least one query ran", len(compiled) > 0)
    check("every query binds entity_id as a parameter",
          all(params.get("entity_id") == eid for eid, params, _ in compiled),
          str([(e, p.get("entity_id")) for e, p, _ in compiled if p.get("entity_id") != e][:2]))
    check("no query inlines the entity id into the SQL",
          all(eid not in sql for eid, _, sql in compiled))
    check("every query filters on entity_id",
        all("{entity_id:String}" in sql for _, _, sql in compiled))
    check("only entities A and B were ever queried",
          {eid for eid, _, _ in compiled} == {ENTITY_A, ENTITY_B},
          str({eid[:8] for eid, _, _ in compiled}))

    print()
    if failures:
        print(f"FAILURES ({len(failures)}):")
        for f in failures:
            print(f"  - {f}")
        return 1
    print("ENTITY_SCOPE_PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/verify/entity_scope.mjs
// G27: a conversation is locked to one entity, entity ids never leave the API in the clear, and
// nothing is answered before an entity is chosen. Every figure is recomputed from data/raw/*.csv.
import fs from 'node:fs';
import path from 'node:path';
import { API, ROOT, fetchRetry, loadAccounts, loadTransactions, defaultEntity, lastMonthOf,
         sumWhere, entityTokenFor, pass, fail } from './_lib.mjs';

const GATE = 'ENTITY_SCOPE';
const txns = loadTransactions();
const entity = defaultEntity(txns);
const prev = lastMonthOf(txns);
const mine = txns.filter(r => r.entity_id === entity);

const accountsCsv = fs.readFileSync(path.join(ROOT, 'data/raw/account.csv'), 'utf8');
const rawEntities = [...new Set(loadAccounts().map(a => a.entity_id).filter(Boolean))];
if (rawEntities.length < 2) fail(GATE, 'account.csv has fewer than two entities; scoping proves nothing');

// Every body the flow produces is swept for a raw entity id at the end.
const bodies = [];
async function chat(body, label) {
  const res = await fetchRetry(`${API}/api/v1/chat`, {
    method: 'POST', headers: { 'content-type': 'application/json' }, body: JSON.stringify(body),
  });
  const text = await res.text();
  bodies.push({ label, text });
  let parsed = null;
  try { parsed = JSON.parse(text); } catch { fail(GATE, `${label}: non-JSON body ${text.slice(0, 200)}`); }
  return parsed;
}

function findRawEntity(text) {
  return rawEntities.filter(e => text.includes(e));
}

// --- (a) no entity chosen: ask, with masked labels and opaque values -------------------------
const ask = await chat({ message: 'what is my balance?', conversation_id: `gate-entity-a-${Date.now()}` }, 'no entity');
if (ask.state !== 'clarification_required') fail(GATE, `no-entity state=${ask.state} (${ask.message || ''})`);
if (ask.clarification?.field !== 'entity') fail(GATE, `no-entity field=${ask.clarification?.field}`);
if (ask.answer || ask.evidence) fail(GATE, 'the entity clarification carried an answer or evidence');
const options = ask.clarification.options || [];
if (options.length < 2) fail(GATE, `entity clarification offered ${options.length} options`);
for (const o of options) {
  if (!/^\*+[0-9a-f]{4}$/.test(o.label)) fail(GATE, `entity option label is not masked: "${o.label}"`);
  if (!o.value || typeof o.value !== 'string') fail(GATE, `entity option has no value: ${JSON.stringify(o)}`);
  if (accountsCsv.includes(o.value)) fail(GATE, `entity option value "${o.value}" appears verbatim in account.csv`);
  if (rawEntities.includes(o.value)) fail(GATE, `entity option value is a raw uuid: ${o.value}`);
}
const labels = options.map(o => o.label);
const mask = e => '*'.repeat(e.length - 4) + e.slice(-4);
const known = new Set(rawEntities.map(mask));
for (const l of labels) if (!known.has(l)) fail(GATE, `offered label ${l} matches no entity in account.csv`);
if (!labels.includes(mask(entity)))
  fail(GATE, `the default entity is not offered as its masked form ${mask(entity)}`);

// --- (b) answering with a token answers the question ------------------------------------------
const token = await entityTokenFor();
const other = await entityTokenFor('other');
if (!other) fail(GATE, 'could not obtain a second entity token');
if (other === token) fail(GATE, 'the "other" token is the same string as the default token');

const conv = `gate-entity-${Date.now()}`;
const CP = 'SWIGGY INSTAMART';
const expected = sumWhere(mine, r => r.txn_date.startsWith(prev) && r.counterparty === CP
                                  && r.transaction_type === 'debit');
if (!expected.count) fail(GATE, `fixture has no ${CP} debits in ${prev}`);

const answered = await chat({ message: `How much did I spend with ${CP} last month?`,
                              entity_id: token, conversation_id: conv }, 'first token');
if (answered.state !== 'answer') fail(GATE, `token turn state=${answered.state} (${answered.message || ''})`);
const total = answered.evidence?.facts.find(f => f.key === 'total');
if (!total) fail(GATE, 'token turn produced no total fact');
if (Math.abs(Number(total.value) - expected.total) > 0.02)
  fail(GATE, `token turn figure ${total.value} != independent ${expected.total}`);

// --- (c) a different token on the same conversation is refused --------------------------------
const SWITCH = "I don't have any Idea what you're talking about.";
const switched = await chat({ message: `How much did I spend with ${CP} last month?`,
                              entity_id: other, conversation_id: conv }, 'switched token');
if (switched.state !== 'out_of_scope') fail(GATE, `entity switch state=${switched.state}, expected out_of_scope`);
if (!(switched.message || '').startsWith(SWITCH))
  fail(GATE, `entity switch message does not start with the fixed refusal: ${JSON.stringify((switched.message || '').slice(0, 120))}`);
if (switched.answer) fail(GATE, 'the refused switch carried an answer');
if (switched.evidence) fail(GATE, 'the refused switch carried evidence');
if (switched.plan) fail(GATE, 'the refused switch carried a plan');

// --- (d) the original token still answers -----------------------------------------------------
const CP2 = 'ZOMATO';
const expected2 = sumWhere(mine, r => r.txn_date.startsWith(prev) && r.counterparty === CP2
                                   && r.transaction_type === 'debit');
if (!expected2.count) fail(GATE, `fixture has no ${CP2} debits in ${prev}`);
const back = await chat({ message: `How much did I spend with ${CP2} last month?`,
                          entity_id: token, conversation_id: conv }, 'original token again');
if (back.state !== 'answer') fail(GATE, `same-token turn state=${back.state} (${back.message || ''})`);
const total2 = back.evidence?.facts.find(f => f.key === 'total');
if (!total2) fail(GATE, 'same-token turn produced no total fact');
if (Math.abs(Number(total2.value) - expected2.total) > 0.02)
  fail(GATE, `same-token figure ${total2.value} != independent ${expected2.total}`);

// --- (e) a tampered or garbage token never answers --------------------------------------------
const tampered = token.slice(0, -6) + (token.slice(-6, -5) === 'A' ? 'B' : 'A') + token.slice(-5);
const badTokens = [
  ['garbage token', 'not-a-real-entity-token'],
  ['tampered token', tampered],
  ['raw uuid as token', rawEntities[0]],
];
const badStates = [];
for (const [label, value] of badTokens) {
  const r = await chat({ message: 'what is my balance?', entity_id: value,
                         conversation_id: `gate-entity-bad-${label.replace(/\W/g, '')}-${Date.now()}` }, label);
  if (r.state === 'answer') fail(GATE, `${label} produced an answer: ${r.answer}`);
  if (r.evidence) fail(GATE, `${label} carried evidence`);
  badStates.push(`${label} -> ${r.state}`);
}

// --- (f) no raw entity uuid appears in any body, and the checker can see one -------------------
const leaked = [];
for (const b of bodies) for (const e of findRawEntity(b.text)) leaked.push(`${b.label}: ${e}`);
if (leaked.length) fail(GATE, `raw entity id in a response body -> ${[...new Set(leaked)].slice(0, 3).join('; ')}`);
const control = findRawEntity(JSON.stringify({ state: 'answer', evidence: { entities_resolved: { entity_id: rawEntities[0] } } }));
if (!control.length) fail(GATE, 'positive control passed: the checker cannot see a raw entity id');

pass(GATE,
  `no entity -> clarification field=entity, ${options.length} options, all labels masked (${labels[0]}) and no value in account.csv`,
  `token -> ${answered.answer} (independent ${expected.total} over ${expected.count} debits)`,
  `different token on the same conversation -> out_of_scope, fixed refusal, no answer/evidence/plan`,
  `original token again -> ${back.answer} (independent ${expected2.total} over ${expected2.count} debits)`,
  `rejected tokens: ${badStates.join(' | ')}`,
  `${bodies.length} bodies swept, none carried any of the ${rawEntities.length} raw entity ids; control caught ${control.length}`);
